### DISCUSSION TITLE
DuckDB with cloud storage?!

### DISCUSSION THREAD

**Discussion by gbusch** [duckdb/duckdb#4646] [Q&A]
I recently came across DuckDB and find the performance and simplicity quite impressive.

Now my question is: what are best practices to use it with cloud storage or is this not a valid use case?

Concretely: We have some parquet / delta tables in an Azure Data Lake Storage and want to display it in Apache Superset. At the moment, we use a Databricks Spark Cluster to query the data which - in my opinion - is absolute overkill.

Are there best practices for using DuckDB in this use case? 
I did some experiments but it was rather slow...
I am afraid that the power of DuckDB is only when data and compute are at the same location which would not be the case here, since the data changes regularly.

**Comment by Alex-Monahan** [comment]
Howdy! Would you mind sharing some more details about what you have tried so far?

**Comment by gbusch** [comment]
Hi Alex,
I essentially followed this post: https://github.com/duckdb/duckdb/discussions/2565#discussioncomment-2844606

My test data is from the Kaggle 2009-2018 airlines set. I copied the data to one parquet file using duckDB and then stored the parquet (~3GB) in an Azure Storage Account. I did not yet try to partition the data.

I then access the data with
```
%sql duckdb:////Users/gerold/OpenSource/duckdb-demo/my-db.duckdb

from adlfs import AzureBlobFileSystem
import pyarrow.dataset as ds

account_name = <account_name>
account_key = <account_key>
abfs = AzureBlobFileSystem( account_name = account_name, account_key = account_key, container_name = "airlines")
pqdata = ds.dataset("airlines/data.parquet", filesystem=abfs, format="parquet")
```

and made the following query:
```
%%sql

SELECT OP_CARRIER, MAX(ARR_DELAY) FROM pqdata GROUP BY OP_CARRIER;
```

This query takes 1.2s locally, but almost 3 minutes from the Azure account.
I guess, this is a network issue...
What worries me most is that when I execute the query again, it is the same duration, so there is no caching.
Is this maybe a configuration issue?

**Reply by arjenpdevries** [reply]
You can create a temporary table, that would cache the results between multiple queries!

As in: https://github.com/duckdb/duckdb/issues/4180#issuecomment-1191016741

**Reply by djouallah** [reply]
@gbusch where did you run your code, I hope it is not in your local laptop ? ideally python code should be run in the same region as your azure storage

**Comment by gbusch** [comment]
First of all, thanks a lot for the quick and friendly feedback in this forum! I am sorry that I could not follow up earlier.

I tried out a few more things but still have some open questions.

* Caching sounds reasonable in the case where I have some related queries but might be a problem when the original data changes.
* Running the code on an Azure VM in the same region speeds up the queries significantly, but still some queries take much longer than expected. 
For example "count(*)" for a table takes some minutes (better than from the local laptop where it takes >30min). However, when accessing a local parquet it is incredibly fast. Also - and that is interesting to me - when using DuckDB's httpfs-extension for S3, it is pretty fast. 
Do I understand it correctly that the local parquet-reader and the httpfs-extension access parquet-meta-data directly, while my Azure-Connection first needs to convert to an Arrow-Table which costs a lot of performance?
* Is there any plan to extend the httpfs-extension to other Cloud-Providers than AWS?

**Comment by djouallah** [comment]
I have spent some time trying to figure out a good worklow, my personal take, copy data to your local ssd in VM , notebook etc and run Queries from there, Cloud storage is simply too slow, it does not matter if it is AWS/ Azure or GCP
you may find this example useful
https://colab.research.google.com/drive/13brUCfUzxh5PVY9OhiIAKjHEdWK0AA3r
